video/out/opengl/lcms.c

```diff
@@ -166,16 +168,88 @@ void gl_lcms_set_memory_profile(struct gl_lcms *p, bstr *profile)
     p->icc_size = profile->len;
 }
 
-// Return and _reset_ whether the lookul table has changed since the last call.
-// If it has changed, gl_lcms_get_lut3d() should be called.
-bool gl_lcms_has_changed(struct gl_lcms *p)
+// Return and _reset_ whether the profile or config has changed since the last
+// call. If it has changed, gl_lcms_get_lut3d() should be called.
+bool gl_lcms_has_changed(struct gl_lcms *p, enum mp_csp_prim prim,
+                         enum mp_csp_trc trc)
 {
-    bool change = p->changed;
+    bool change = p->changed || p->prev_prim != prim || p->prev_trc != trc;
     p->changed = false;
+    p->prev_prim = prim;
+    p->prev_trc = trc;
     return change;
 }
 
-bool gl_lcms_get_lut3d(struct gl_lcms *p, struct lut3d **result_lut3d)
+static cmsHPROFILE get_vid_profile(cmsContext cms, cmsHPROFILE disp_profile,
+                                   enum mp_csp_prim prim, enum mp_csp_trc trc)
+{
+    // The input profile for the transformation is dependent on the video
+    // primaries and transfer characteristics
+    struct mp_csp_primaries csp = mp_get_csp_primaries(prim);
+    cmsCIExyY wp_xyY = {csp.white.x, csp.white.y, 1.0};
+    cmsCIExyYTRIPLE prim_xyY = {
+        .Red   = {csp.red.x,   csp.red.y,   1.0},
+        .Green = {csp.green.x, csp.green.y, 1.0},
+        .Blue  = {csp.blue.x,  csp.blue.y,  1.0},
+    };
+
+    cmsToneCurve *tonecurve = NULL;
+    switch (trc) {
+    case MP_CSP_TRC_LINEAR:  tonecurve = cmsBuildGamma(cms, 1.0); break;
+    case MP_CSP_TRC_GAMMA18: tonecurve = cmsBuildGamma(cms, 1.8); break;
+    case MP_CSP_TRC_GAMMA22: tonecurve = cmsBuildGamma(cms, 2.2); break;
+    case MP_CSP_TRC_GAMMA28: tonecurve = cmsBuildGamma(cms, 2.8); break;
+
+    case MP_CSP_TRC_SRGB:
+        // Values copied from Little-CMS
+        tonecurve = cmsBuildParametricToneCurve(cms, 4,
+                (double[5]){2.40, 1/1.055, 0.055/1.055, 1/12.92, 0.04045});
+        break;
+
+    case MP_CSP_TRC_PRO_PHOTO:
+        tonecurve = cmsBuildParametricToneCurve(cms, 4,
+                (double[5]){1.8, 1.0, 0.0, 1/16.0, 0.03125});
+        break;
+
+    case MP_CSP_TRC_BT_1886: {
+        // To build an appropriate BT.1886 transformation we need access to
+        // the display's black point, so we use the reverse mappings
+        cmsHPROFILE xyz_profile = cmsCreateXYZProfileTHR(cms);
+        cmsHTRANSFORM rgb2xyz = cmsCreateTransformTHR(cms,
+                disp_profile, TYPE_RGB_16, xyz_profile, TYPE_XYZ_DBL,
+                INTENT_RELATIVE_COLORIMETRIC, 0);
+        cmsCloseProfile(xyz_profile);
+        if (!rgb2xyz)
+            return false;
+
+        uint64_t black[3] = {0};
+        cmsCIEXYZ disp_black;
+        cmsDoTransform(rgb2xyz, black, &disp_black, 1);
+
+        // Build the parametric BT.1886 transfer curve
+        const double gamma = 2.40;
+        double binv = pow(disp_black.Y, 1.0/gamma);
+        double gain = 1.0 - binv;
+        double lift = binv / gain;
```

Comment:
Just double checking:

The LittleCMS formula is `Y = (a·X + b)^g + c`. The BT.1886 formula is `L = gain·(V+lift)^g` with `gain = (Lw_inv - Lb_inv)^g` and `lift = Lb_inv / (Lw_inv - Lb_inv)`.

Fitting this into the LittleCMS formula gives us:

```
(a·X + b)^g = gain · (V+lift)^g
(a·X + b)^g = (Lw_inv - Lb_inv)^g · (V+lift)^g
(a·X + b)^g = ((Lw_inv - Lb_inv) · V + (Lw_inv - Lb_inv) · lift)^g
⇒ a = Lw_inv - Lb_inv = 1.0 - binv
  b = (Lw_inv - Lb_inv) · lift = (Lw_inv - Lb_inv) · Lb_inv / (Lw_inv - Lb_inv) = Lb_inv
```

Looks like you were right. I missed the distribution over the addition inside the right hand side when doing this in my head.